File: Pythoncode/22.括号生成.py
class Solution:
    """
    https://leetcode-cn.com/problems/generate-parentheses/solution/hui-su-suan-fa-by-liweiwei1419/
    """
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        leftnum = n
        rightnum = n
        self.helper(leftnum, rightnum, "", res)
        return res
    
    def helper(self, leftnum, rightnum, s, res):
        if leftnum == 0 and rightnum == 0:
            res.append(s)
        if leftnum != 0:
            copys = s[:]
            copys += "("
            self.helper(leftnum-1, rightnum, copys, res)
        if rightnum - 1 >= leftnum:
            copys = s[:]
            copys += ")"
            self.helper(leftnum, rightnum-1, copys, res)

File: Cppcode/22.括号生成.cpp
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        if (n == 0) return res;
        helper(n, n, res, "");
        return res;
    }

    void helper(int left_bracket_num, int right_bracket_num, vector<string> &res, string cur_string)
    {
        if (left_bracket_num == 0 and right_bracket_num == 0) res.push_back(cur_string);
        if (left_bracket_num > 0) helper(left_bracket_num-1, right_bracket_num, res, cur_string+'(');
        if (right_bracket_num > 0 and right_bracket_num - 1 >= left_bracket_num) helper(left_bracket_num, right_bracket_num-1, res, cur_string+")");
    }
};